presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlockBuilder.java

```diff
@@ -504,6 +504,19 @@ private static int getHashPosition(Block keyBlock, int position, MethodHandle ke
             throw new RuntimeException(throwable);
         }
 
-        return (int) Math.floorMod(hashCode, hashTableSize);
+        return getReducedHashPosition(hashCode, hashTableSize);
+    }
+
+    // Adapted from the fastrange algorithm: https://github.com/lemire/fastrange. This function reduces the hashcode
```

Comment:
Since the link contains a quite comprehensive documentation about this algorithm and the proof, I would just say:

```
    // Adapted from the fastrange algorithm: https://github.com/lemire/fastrange.
    // We first transform the hashcode to 32bit integer, and then apply the fastrange algorithm to
    // map the hashcode to [0, hashTableSize)
```

What do you think, @electrum , @haozhun ?

Reply:
That's ok, but I don't know why we're making this more complicated than it really is. The core of this method is just doing a simple normalization of values from one range to another using standard math. I.e:

```
y = (x - x0) / (x1 - x0) * (y1 - y0) + y0

where 
   x0 = 0
   x1 = 2^32
   y0 = 0
   y1 = hash size
```

so,
```
y = x / 2^32 * hashSize
```

Of course, instead of dividing by 2^32, it performs that operation with a shift (btw, it's possible the VM might optimize this, so expressing it that way could be more intuitive and as fast).

Not sure why it's being called the "fastrange algorithm".

Reply:
Yes totally agree. It is just doing normalization. I will remove the extra explanations. But division is very expensive and JVM won't be able to optimize it because it doesn't know the divisor is 2^32. The signed 64 bit divide on Intel CPU takes a few dozens of cycles. The floating point division is cheaper but still slow. This can be shown in additional benchmarks:
Benchmark                              Mode  Cnt   Score   Error  Units
BenchmarkFloorMod.ReduceHashPosition6  avgt   10   5.186 ± 0.306  ms/op
BenchmarkFloorMod.ReduceHashPosition7  avgt   10  10.207 ± 0.453  ms/op
where the definitions are
private int getReducedHashPosition6(long hashcode, int hashTableSize)
    {
        return (int)(((double)(hashcode & 0xffff_ffffL) / powerOf2To32) * hashTableSize);
    }

    private int getReducedHashPosition7(long hashcode, int hashTableSize)
    {
        return (int)((hashcode & 0xffff_ffffL) * hashTableSize / powerOf2To32);
    }

Reply:
> btw, it's possible the VM might optimize this

In fact, it does if it can prove the numerator is positive:

e.g., 

```java
(value & 0x7FFF_FFFFL) * (hashSize & 0x7FFF_FFFFL) / (1L << 32);
```

compiles to:

```
mov    %rbp,0x10(%rsp)
and    $0x7fffffff,%r10d
mov    $0x7fffffff,%eax
and    0x10(%rsi),%rax
imul   %r10,%rax
sar    $0x20,%rax
```

Reply:
@martint Yes it looks like JVM can do this if the divisor is constant (final).  I can change it to the way using divide but I just feel writing in the shifting way looks simpler and is self-explanatory for the performance gain. How about we keep the code using the shift and simplify the comment as 

```
// This function reduces the 64 bit hashcode to [0, hashTableSize) uniformly by first reducing the hashcode to 32 bit 
    // integer then applying normalization on it to reduce the range from [0, 2^32) to [0, hashTableSize)
```